motor-control/firmware/stepper_motor/motor_hardware.cpp

```diff
@@ -4,64 +4,28 @@
 #include "common/firmware/gpio.hpp"
 #include "motor-control/firmware/motor_control_hardware.h"
 
-using namespace motor_hardware;
+namespace motor_hardware {
 
-void MotorHardware::step() { gpio::set(pins.step); }
-
-void MotorHardware::unstep() { gpio::reset(pins.step); }
-
-void MotorHardware::positive_direction() { gpio::set(pins.direction); }
-void MotorHardware::negative_direction() { gpio::reset(pins.direction); }
-void MotorHardware::activate_motor() { gpio::set(pins.enable); }
-void MotorHardware::deactivate_motor() { gpio::reset(pins.enable); }
-void MotorHardware::start_timer_interrupt() {
-    LOG("Starting timer interrupt")
-    motor_hardware_start_timer(tim_handle);
-}
-void MotorHardware::stop_timer_interrupt() {
-    motor_hardware_stop_timer(tim_handle);
-}
-void MotorHardware::read_limit_switch() {
-    limit = (gpio::is_set(pins.limit_switch));
+template <>
+void MotorHardware<HardwareConfig>::activate_motor() {
+    gpio::set(pins.enable);
 }
 
-void MotorHardware::read_estop_in() { estop = gpio::is_set(pins.estop_in); }
-
-void MotorHardware::read_sync_in() { sync = (gpio::is_set(pins.sync_in)); }
-
-void MotorHardware::set_LED(bool status) {
-    if (status) {
-        gpio::set(pins.led);
-    } else {
-        gpio::reset(pins.led);
-    }
+template <>
+void MotorHardware<HardwareConfig>::deactivate_motor() {
```

Comment:
I see what you're going for here, but we really shouldn't have explicit specializations for application-dependent types in library code like this - and having these specializations and _only_ these specializations live in a .cpp file instead of the header makes it harder to read. Given that `MotorHardware` is kind of trivial, i think it's probably better to just have a `HeadMotorHardware` that lives in `include/head/firmware` that does this and can explicitly handle the brake pins, and get these definitions unspecialized and in the header file.

Reply:
okayyyy 😢